modules/RTC/TraceablePeerConnection.js

```diff
@@ -689,12 +1088,294 @@ TraceablePeerConnection.prototype.addStream = function(stream, ssrcInfo) {
+        logger.error(
+            'An attempt to attach an already attached '
+            + `${localTrack} to ${this} was made`);
+
+        return;
+    }
+
+    localTrack._addPeerConnection(this);
+
+    logger.debug(`Attached ${localTrack} to ${this}`);
+
+    // Muted video tracks are not added to the PeerConnection
+    if (!localTrack.isVideoTrack() || !localTrack.isMuted()) {
+        const webRtcStream = localTrack.getOriginalStream();
+
+        if (webRtcStream) {
+            this._addStream(webRtcStream);
+
+            return true;
+        }
+
+        logger.error(`${this} attach - no WebRTC stream for: ${localTrack}`);
+
+        return false;
+    }
+
+    logger.debug(`${this} attach ${localTrack} - not adding to PC`);
+
+    return false;
+};
+
+/**
+ * Remove local track from this TPC.
+ * @param {JitsiLocalTrack} localTrack the track to be removed from this TPC.
+ */
+TraceablePeerConnection.prototype.removeTrack = function(localTrack) {
+    const webRtcStream = localTrack.getOriginalStream();
+
+    this.trace(
+        'removeStream',
+        localTrack.rtcId, webRtcStream ? webRtcStream.id : undefined);
+
+    if (!this._assertTrackBelongs('removeStream', localTrack)) {
+        // Abort - nothing to be done here
+        return;
+    }
+    this.localTracks.delete(localTrack.rtcId);
+    this.localSSRCs.delete(localTrack.rtcId);
+
+    // A detached track will not require removal
+    if (this._isTrackAttached(localTrack)) {
+        localTrack._removePeerConnection(this);
+    }
+
+    if (webRtcStream) {
+        if (RTCBrowserType.isFirefox()) {
+            this._handleFirefoxRemoveStream(webRtcStream);
+        } else {
+            this.peerconnection.removeStream(webRtcStream);
+        }
+    }
+};
+
+/**
+ * Removes local track as part of the unmute operation.
+ * @param {JitsiLocalTrack} localTrack the local track to be remove as part of
+ * the mute operation.
+ * @return {boolean} <tt>true</tt> if the underlying PeerConnection's state has
+ * changed and the renegotiation is required or <tt>false</tt> otherwise.
+ */
+TraceablePeerConnection.prototype.removeTrackMute = function(localTrack) {
+    const webRtcStream = localTrack.getOriginalStream();
+
+    this.trace(
+        'removeStreamMute',
+        localTrack.rtcId, webRtcStream ? webRtcStream.id : null);
+
+    if (!this._assertTrackBelongs('removeStreamMute', localTrack)) {
+        // Abort - nothing to be done here
+        return false;
+    } else if (!localTrack._isAttachedToPC(this)) {
+        // Abort - nothing to be done here
+        logger.warn(
+            `Not removing detached ${localTrack} as unmute from ${this}`);
+
+        return false;
+    }
+
+    if (webRtcStream) {
+        logger.info(
+            `Removing ${localTrack} as mute from ${this}`);
+        this._removeStream(webRtcStream);
+
+        return true;
+    }
+
+    logger.error(`removeStreamMute - no WebRTC stream for ${localTrack}`);
+
+    return false;
```

Comment:
Probably the normal one should also return a boolean. It optimises the case where no change to the peer connection was made, so the SRD/sLD cycle is not necessary. The method will probably go away, because Brian suggested to use detach and I think it makes sense. I'm adding a FIXME for now.